In [47]:
import pandas as pd
from pathlib import Path
import numpy as np
import re
from huggingface_hub import HfApi

In [48]:
#python -m pip install huggingface_hub
#python -m pip install ipywidgets

#open cmd

In [49]:
pd.set_option('display.max_columns', None)

# Option 1:
Read all parquet files from Hugging face directly into one dataframe (no download):

In [50]:
print('Skip this cell')

'''
api = HfApi()
files = api.list_repo_files(
    repo_id="rubiogarciaf/DSCI6523_LeagueofLegends",
    repo_type="dataset"
)
parquet_files = [f for f in files if f.endswith(".parquet")]

base = "https://huggingface.co/datasets/rubiogarciaf/DSCI6523_LeagueofLegends/resolve/main"
df = pd.concat(
    [pd.read_parquet(f"{base}/{f}") for f in parquet_files],
    ignore_index=True
)
df.head(5)
'''

Skip this cell


'\napi = HfApi()\nfiles = api.list_repo_files(\n    repo_id="rubiogarciaf/DSCI6523_LeagueofLegends",\n    repo_type="dataset"\n)\nparquet_files = [f for f in files if f.endswith(".parquet")]\n\nbase = "https://huggingface.co/datasets/rubiogarciaf/DSCI6523_LeagueofLegends/resolve/main"\ndf = pd.concat(\n    [pd.read_parquet(f"{base}/{f}") for f in parquet_files],\n    ignore_index=True\n)\ndf.head(5)\n'

In [51]:
# Skip this one too
'''
df.shape
'''

'\ndf.shape\n'

# Option 2:
Download locally from HuggingFace.

In [52]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [53]:
from huggingface_hub import snapshot_download

local_dir = snapshot_download(
    repo_id="rubiogarciaf/DSCI6523_LeagueofLegends",
    repo_type="dataset",
    local_dir="/content/drive/MyDrive/newman_ml_ops/data_huggingface_new" # Updated to the specified Google Drive path
)
print(f"Downloaded to {local_dir}")

Fetching 23 files:   0%|          | 0/23 [00:00<?, ?it/s]

Downloaded to /content/drive/MyDrive/newman_ml_ops/data_huggingface_new


In [54]:
import glob

full_df = pd.read_parquet(glob.glob(local_dir + "/*.parquet"))
# 20260522 JSP Add a source_row_id index to uniquely identify each row (small change gameId is not unique)
full_df = full_df.reset_index(drop=False).rename(columns={"index": "source_row_id"})
full_df.head(5)

,source_row_id,gameId,gameCreation,gameCreationDate,gameDuration,patch,tier,average_leaguePoints,bwin,b1_puuid,b1_teamPosition,b1_championId,b1_leaguePoints,b1_mpoints,b2_puuid,b2_teamPosition,b2_championId,b2_leaguePoints,b2_mpoints,b3_puuid,b3_teamPosition,b3_championId,b3_leaguePoints,b3_mpoints,b4_puuid,b4_teamPosition,b4_championId,b4_leaguePoints,b4_mpoints,b5_puuid,b5_teamPosition,b5_championId,b5_leaguePoints,b5_mpoints,r1_puuid,r1_teamPosition,r1_championId,r1_leaguePoints,r1_mpoints,r2_puuid,r2_teamPosition,r2_championId,r2_leaguePoints,r2_mpoints,r3_puuid,r3_teamPosition,r3_championId,r3_leaguePoints,r3_mpoints,r4_puuid,r4_teamPosition,r4_championId,r4_leaguePoints,r4_mpoints,r5_puuid,r5_teamPosition,r5_championId,r5_leaguePoints,r5_mpoints
0,0,5554887388,1778112012000,05/07/2026,1628,169,Bronze I,747.4,1,aIo-ehCfVztnixyhx_moFpzMFWCLYxAhDEf1PbUsbh-c_i...,TOP,122,703.0,107496.0,lkq_x8mLEpambpgjL4KZk_GbwK1BwXX6d-FdNbY5DZSfa4...,JUNGLE,33,646.0,2517.0,8Nwul0Nsqoxjiy_sgiC06JaG5HnIHSYQf5VbTbZzu4L5RW...,MIDDLE,131,763.0,434369.0,fgRiQcyC6EWhQ8aVXy4fnRF1lyUopjk3mwwcE8Z1MjVxfv...,BOTTOM,202,603.0,48394.0,nEUyoLbGbON4YmYcFvC-6_5zCQ6tSeyXw_7R2S-nNMkxOd...,UTILITY,57,835.0,12810.0,eqZ6PmWmnUhRTVfU0JB6KsvZ3imVhbTnJR59HBH_NcrvHQ...,TOP,875,677.0,222810.0,_6HcU9kSqlzwamTGeu2DPfDqixm9QBPika_nmXG5-ouA2c...,JUNGLE,11,830.0,27893.0,oVp8tv9o3cfPZUoVeHKDBCYZu7hYWXNtIOSJINt00xIMFX...,MIDDLE,245,818.0,10583.0,QcYwM6qKMADC5r1_blhy8mq0_6NimMo_ddwFOHP9GHzeqB...,BOTTOM,804,901.0,23050.0,8RbPWKWQ5C-GauAK4rIs9TPEc3__JQzW7x7wQGgt8OQNAE...,UTILITY,117,698.0,56880.0
1,1,5554889227,1778112280000,05/07/2026,2073,169,Bronze I,724.8,0,DCxZCI9BawhcsYQEVH1hS6YqVS8jsjqTJiRAXTl-EWYnx8...,TOP,122,801.0,139244.0,ZJ8sZhe0r6v6VVsCM3QRwBMDvx6wEdgqQsF0_JCKgnzpMJ...,JUNGLE,23,705.0,55457.0,b3eabXCPuSlu1iwEJpuv9xj4ObR7KmrPtwgGbMJagbpT6r...,MIDDLE,143,628.0,300018.0,gH-K1Aieq_BUaDfk_mXD66S_VBn4mXTYjNZV3KYA0oCyWy...,BOTTOM,901,626.0,33846.0,kk8V_zy2r68viHpT9ZQrdI7ZFjuRYHZCLPNsDPlwEWPLSp...,UTILITY,147,788.0,2196.0,TvzhWJjVdvd6R2TumwXBKa4gNTvlDZ_roMfG86RXRNCIwd...,TOP,86,669.0,451610.0,KZXoT0yK3qHq7qRyCICnkjccxMDx9FQSBBAXLroJpMd7UI...,JUNGLE,233,586.0,58499.0,isFJmmI1kpuqnrj6BhpY2piyzmAoiyPMu4yy-9SbvpzKH6...,MIDDLE,7,920.0,67763.0,et3CI2OBH2SyFGoPBFl508ashVgZWFnCbQHmIFcKS4FoGs...,BOTTOM,51,608.0,198145.0,j2E6If153AMjHwAy6oAiNWmwkrtS5eEfJEXqVIiJVfr4U8...,UTILITY,902,917.0,1305.0
2,2,5554889599,1778112328000,05/07/2026,1865,169,Bronze I,725.7,0,y2f--nrfSidXf2XE_nzu2NUGuiwp-COxga6J6Jxsf1MAZa...,TOP,875,565.0,54369.0,k8M0WpMrr0Ju213ZlxP0quWl5Uv4YC7JML2WbkEaAmYzgX...,JUNGLE,141,818.0,10416.0,WgVmQdiY3wvdxQ1dEJIgzPDtLPerMbBxBcwS8QQxMdAqhM...,MIDDLE,202,1065.0,32350.0,TniIhqRqj9xsS1pqtHdmnvtfmiSw9Hzi2Wsd7jJDR7WWQ3...,BOTTOM,63,569.0,40707.0,xOkpUQLPAEMS3kkCuCeaaMDGbmwm7T4FORyXuct6upT7vz...,UTILITY,22,493.0,618.0,0pFb0Tu4gMEM4OPBnrNasfLZ0rOzVWPeR7qUuRbjS4D0FT...,TOP,68,776.0,584241.0,J-Exg1rLEYyfgOcEfIo83otGQtnvjkFQIY3BFaF025-yC9...,JUNGLE,11,823.0,192565.0,7MeLyJ9bYHQhm_COoCjR9eXg8AQgW-RgMIlBGYtxpemY5L...,MIDDLE,50,648.0,252752.0,xLK9nrYIjkJw3G3QjMtoB-VOX6eSi711IKQNVC7-jpv-cM...,BOTTOM,51,671.0,25955.0,QPu2gO5XbyNIvnrTRdoYgG7Tioz6C__ICrvs4sVYy4yGua...,UTILITY,17,829.0,1860657.0
3,3,5554891355,1778112259000,05/07/2026,2476,169,Bronze I,789.6,1,eEW6Y0mUP7dMcbh-Nk7HfpQWg1RvTTrDfTa0YxhYFdQg3I...,TOP,516,848.0,135.0,DA4j-3TcryEwVV9mROZ9Wvfd0MOBVBOox_LAxqpSxPpnn1...,JUNGLE,35,800.0,23831.0,c-_bRHyJbZlBCUss7f1O088rn75h6bUkYlv52yrK-smPPD...,MIDDLE,777,509.0,25527.0,cufryKbdY8ikPboaAMo6ybPfJiBXsIIO_pW5xJ8KWEb9KT...,BOTTOM,51,893.0,581416.0,wEDnzTqariqDUTqshpqVT7GnxO6-ITfpVfis4GkTT9mCpT...,UTILITY,26,811.0,662.0,625BDKUYC_7NuVIBsO96gP-72QzVbhPLUmIPYSU7H4_qkO...,TOP,875,853.0,118720.0,z6wsfTF8dbzUJnqUq6YG2arjdN6FipOEQpdNgKqbNyKc2N...,JUNGLE,102,701.0,1497.0,S1y-mk9h5B7HVile7YrugeSieGWxISuoikTOEk3f0PYMSC...,MIDDLE,266,890.0,21243.0,SSsJyr19SnVu_ysQsvbikTWqBAKIYlsdoUnpFNGNia-m_R...,BOTTOM,360,771.0,298294.0,hRttrpbPj9bOwupzVhfLhjoEby0GgGB05Ap-ajBMinvT_d...,UTILITY,202,820.0,7975.0
4,4

In [55]:
full_df.shape

(552254, 59)

In [56]:
429651 * 10

4296510

In [57]:
full_df.groupby('gameCreationDate').count()['gameId']

,gameId
gameCreationDate,
05/01/2026,16572
05/02/2026,18270
05/03/2026,19513
05/04/2026,20585
05/05/2026,23271
05/06/2026,26225
05/07/2026,23140
05/08/2026,21775
05/09/2026,26469


# Assignment 1.1 — Organizational Framing

Before any code, the first assignment asks you to take a position on what this model would actually be for. You've seen the data. You've seen the candidate deployments — matchmaking quality assurance, broadcast win probability, a coaching app, a betting platform, others you might propose. The assignment is to commit to one and reason through what success and failure would mean for that specific deployment.

Full prompt and submission instructions are on Canvas. The short version: pick a deployment, write 700-1,000 words on the use case, stakeholders, failure costs, failure scenarios, and what an acceptable model would look like. No code, no notebook — this one is a written argument.

This is going to feel underdeveloped on your first pass, and that's the point. In week 8 you'll come back to this exercise with seven weeks of technical depth behind you, and the gap between the two attempts is part of what the course is for.

# Assignment 1.2 — Wide to Thin Reshape

Here's the structural problem the data has. Every feature you'll build in this course — recent win rates, LP trajectories, champion pools, role consistency, tilt signals — lives at the player level. But the data arrives at the game level, one row per game, with each of the 10 players encoded as a set of prefix columns (b1_, b2_, ..., r5_).
You can't engineer player-level features from a player-level history if every player is smeared across ten columns in one row. The first piece of infrastructure this course needs is a reshape: from one row per game to ten rows per game, one per player.


- You can use an AI assistant for this. I'm telling you to. The point isn't to memorize pd.wide_to_long or pd.melt. The point is to ask good questions, verify the answers, and produce correct code. That's what working as a data scientist looks like in 2026.
- There's a subtlety in the outcome variable. The dataset has bwin = 1 if the blue team won. When you attach an outcome to each player row, the naive answer is wrong for half your rows. Think before you write df['win'] = df['bwin'].
- Verify your work. Your wide table has N games. Your thin table must have exactly 10N rows. And the row count is necessary but not sufficient — what else has to be true if your reshape worked?

Documentation matters here. Every step should be explained in a text cell. Part 3 of the assignment is a data dictionary of the thin table you produce.

## pd.melt vs. pd.wide_to_long
pd.wide_to_long() works best when columns share a stub name and differ by a suffix, such as score_1, score_2, score_3. This dataset uses prefixes like b1_puuid, r5_championId, so using wide_to_long() would require renaming columns first. pd.melt() is more flexible because it can turn arbitrary columns into key/value rows, but for this dataset it creates a very large intermediate table and then requires another pivot. I ultimately used an explicit slot-loop approach because it preserves column structure directly and avoids the memory overhead of melt.

I'm using the Hermes agent framework backed by gpt-5.5 for AI assistance. After some discussion our conclusion is that pd.melt() is the better option. pd.wide_to_long is a good function but it works more naturally when the repeated-observation identifier appears as a suffix after a shared variable stub. Our columns have the identifier at the end of the prefix. We could rename the columns but that's an extra step and we could just use melt. This seems reasonable and there is very little overhead to implement the more complex melt syntax if AI assisted.

## bwin vs. win
In the original wide data we are provided with bwin which indicates if bwin won the respective match. This is appropriate for the wide structure because each row contains attributes and the result for both blue and red teams. An rwin column would work as well but having both would be redundant as one is the exact inverse of the other.

When we reverse pivot the data and have one record per player we could have retained bwin but then we'd have to carry computational logic for calculating whether a player won based on their team membership throughout the project. It's logical that we'd want to eliminate this complexity and just have a "win" value. To do this we take the bwin value as is for blue team members and inverse (or use 1-bwin) for the red team.

## Column changes
For the narrow format we will (by definition) not have as many columns. The bx and rx columns are reverse pivoted into one player attribute set per record. bwin is transformed into win using rules based on the player team.

I also considered moving the match level attributes to a second target table. This would have eliminated a lot of data duplication through normalization. In the end I decided that the simplicity of a single "wide" narrow table was preferrable to having to carry join logic and compute overhead through the rest of the project.

## Reshape implementation
I originally had this coded up using melt but it wasn't great. Due to the number of records and columns memory consumption blew up colab so I had to add a loop and do the pivot in chunks. This overhead more than exceeded any efficiencies from using melt so I changed strategies.

For my implementation I chose the none of the above option. I experimented with melt but it required a melt which created a key/value pair dataframe (a huge one) that was then reverse pivoted back into the desired shape. This was slow. My implementation below uses looping to work through the columns and create recods of the right shape. These records are then appended to the dataframe.

There are some additional fields that are somewhat duplicative. In particular bwin/win and side_code/side_slot/player_slot don't add data but they might be useful for validation or analysis. I'll leave them in for now.

Note that I have a bit of a weird return structure. I needed the match_cols list in order to validate the values in those columns later.

In [58]:
def make_player_match_df(full_df):
    """Convert full_df from one row per match to one row per player per match.

    The source dataframe has one row per match and player columns named like:
        b1_puuid, b1_teamPosition, ..., r5_mpoints

    The returned dataframe has ten rows per match: one row for each blue/red
    player slot, with all match-level attributes retained.
    """
    player_col_pattern = re.compile(r"^[br][1-5]_")
    player_cols = [
        col for col in full_df.columns
        if player_col_pattern.match(col)
    ]
    match_cols = [
        col for col in full_df.columns
        if col not in player_cols
    ]

    if not player_cols:
        raise ValueError("No player columns found. Expected names like b1_puuid or r5_championId.")

    slot_specs = [
        ("b", "blue", 1),
        ("b", "blue", 2),
        ("b", "blue", 3),
        ("b", "blue", 4),
        ("b", "blue", 5),
        ("r", "red", 1),
        ("r", "red", 2),
        ("r", "red", 3),
        ("r", "red", 4),
        ("r", "red", 5),
    ]

    player_frames = []
    for side_code, team, side_slot in slot_specs:
        player_slot = f"{side_code}{side_slot}"
        prefix = f"{player_slot}_"
        slot_cols = [
            col for col in player_cols
            if col.startswith(prefix)
        ]

        if not slot_cols:
            raise ValueError(f"No columns found for player slot {player_slot}.")

        slot_df = full_df[match_cols + slot_cols].copy()
        rename_map = {
            col: col.removeprefix(prefix)
            for col in slot_cols
        }
        slot_df = slot_df.rename(columns=rename_map)

        insert_at = len(match_cols)
        slot_df.insert(insert_at, "team", team)
        slot_df.insert(insert_at + 1, "side_code", side_code)
        slot_df.insert(insert_at + 2, "side_slot", side_slot)
        slot_df.insert(insert_at + 3, "player_slot", player_slot)

        player_frames.append(slot_df)

    player_match_df = pd.concat(player_frames, ignore_index=True)

    # bwin is 1 when blue won; win is 1 when this player's team won.
    team_is_blue = player_match_df["team"].eq("blue").to_numpy()
    bwin = player_match_df["bwin"].to_numpy(dtype=np.int8, copy=False)

    player_match_df["win"] = np.where(team_is_blue, bwin, 1 - bwin).astype(np.int8)

    # Convert date strings to pandas datetime for efficient date operations.
    if "gameCreationDate" in player_match_df.columns:
        player_match_df["gameCreationDate"] = pd.to_datetime(
            player_match_df["gameCreationDate"],
            format="%m/%d/%Y"
        )

    # Convert whole-number columns to smaller integer dtypes after the reshape.
    # Use nullable integer dtypes for source columns because they may contain missing values.
    integer_dtype_map = {
        "leaguePoints": "Int16",
        "championId": "Int16",
        "mpoints": "Int32",
        "side_slot": "int8",
        "bwin": "int8",
    }

    for col, dtype in integer_dtype_map.items():
        if col in player_match_df.columns:
            player_match_df[col] = player_match_df[col].astype(dtype)

    return player_match_df, match_cols


In [59]:
# Run after full_df is loaded:
player_match_df, match_cols = make_player_match_df(full_df)
print(f"full_df shape: {full_df.shape}")
print(f"player_match_df shape: {player_match_df.shape}")
player_match_df.head(10)

full_df shape: (552254, 59)
player_match_df shape: (5522540, 19)


,source_row_id,gameId,gameCreation,gameCreationDate,gameDuration,patch,tier,average_leaguePoints,bwin,team,side_code,side_slot,player_slot,puuid,teamPosition,championId,leaguePoints,mpoints,win
0,0,5554887388,1778112012000,2026-05-07,1628,169,Bronze I,747.4,1,blue,b,1,b1,aIo-ehCfVztnixyhx_moFpzMFWCLYxAhDEf1PbUsbh-c_i...,TOP,122,703,107496,1
1,1,5554889227,1778112280000,2026-05-07,2073,169,Bronze I,724.8,0,blue,b,1,b1,DCxZCI9BawhcsYQEVH1hS6YqVS8jsjqTJiRAXTl-EWYnx8...,TOP,122,801,139244,0
2,2,5554889599,1778112328000,2026-05-07,1865,169,Bronze I,725.7,0,blue,b,1,b1,y2f--nrfSidXf2XE_nzu2NUGuiwp-COxga6J6Jxsf1MAZa...,TOP,875,565,54369,0
3,3,5554891355,1778112259000,2026-05-07,2476,169,Bronze I,789.6,1,blue,b,1,b1,eEW6Y0mUP7dMcbh-Nk7HfpQWg1RvTTrDfTa0YxhYFdQg3I...,TOP,516,848,135,1
4,4,5554893781,1778112396000,2026-05-07,1905,169,Bronze I,773.9,1,blue,b,1,b1,RKeHIRhaIvutH1wlzHYH8DS2rftsYynj0ifkW9tbCud67r...,TOP,86,947,131860,1
5,5,5554895349,1778112606000,2026-05-07,1839,169,Bronze I,724.1,1,blue,b,1,b1,GEVhi_ZcyeRTSvBdPeK9m88SrO-addSRe6_GbKUlKwbhdA...,TOP,36,819,74284,1
6,6,5554895704,1778112848000,2026-05-07,1894,169,Bronze I,799.4,1,blue,b,1,b1,0IMBgCkxkqMTGtZo5cwVRND5OSzh-UXzNjHKJc_lRzBhLl...,TOP,17,889,377529,1
7,7,5554896796,1778112536000,2026-05-07,1732,169,Bronze I,793.0,0,blue,b,1,b1,O1nsDsEQpQnhObZBS48NeupwjveIgUA4_rRZri22IxCSZ7...,TOP,86,800,138455,0
8,8,5554896813,1778112560000,2026-05-07,1775,169,Bronze I,763.6,1,blue,b,1,b1,gwUni56neeYhX0VofuZfQhuMaiLlhoXi__fQx5RHn3cfZb...,TOP,54,837,8756,1
9,9,5554898560,1778112978000,2026-05-07,1814,169,Bronze I,768.8,1,blue,b,1,b1,tDiHuS4KJ1IClms6LI0eiFWO3v-PrYnSi6dvxH507doVpg...,TOP,875,763,2617353,1


## Reshape validation

In [60]:
# Validation checks for make_player_match_df
expected_slots = {"b1", "b2", "b3", "b4", "b5", "r1", "r2", "r3", "r4", "r5"}

# Test 1: each match should become exactly ten player rows.
assert len(player_match_df) == len(full_df) * 10

# Test 2: no game should lose or gain players during the reshape.
players_per_game = player_match_df.groupby("gameId").size()
assert players_per_game.eq(10).all()

# Test 3: each game should have exactly one row for each expected player slot.
slots_per_game = player_match_df.groupby("gameId")["player_slot"].agg(set)
assert slots_per_game.apply(lambda slots: slots == expected_slots).all()

# Test 4: gameId + player_slot should uniquely identify a player row.
assert not player_match_df.duplicated(["gameId", "player_slot"]).any()

# Test 5: each game should have five blue rows and five red rows.
team_counts = (
    player_match_df
    .groupby(["gameId", "team"])
    .size()
    .unstack(fill_value=0)
    .reindex(columns=["blue", "red"], fill_value=0)
)
assert team_counts["blue"].eq(5).all()
assert team_counts["red"].eq(5).all()

# Test 6: blue player outcomes should equal bwin; red outcomes should be the inverse.
if "bwin" in player_match_df.columns:
    blue_rows = player_match_df["team"].eq("blue")
    red_rows = player_match_df["team"].eq("red")
    assert (player_match_df.loc[blue_rows, "win"] == player_match_df.loc[blue_rows, "bwin"]).all()
    assert (player_match_df.loc[red_rows, "win"] == 1 - player_match_df.loc[red_rows, "bwin"]).all()

# Test 7: whole-number columns should use compact integer dtypes.
expected_dtypes = {
    "leaguePoints": "Int16",
    "championId": "Int16",
    "mpoints": "Int32",
    "side_slot": "int8",
    "bwin": "int8",
    "win": "int8",
    "gameCreationDate": "datetime64[ns]"
}
actual_dtypes = player_match_df[list(expected_dtypes)].dtypes.astype(str).to_dict()
assert actual_dtypes == expected_dtypes, actual_dtypes

# Test 8: validate that the metadata came over properly (unique for each source record)
metadata_cols = [
    col for col in match_cols
    if col in player_match_df.columns
]
metadata_nunique = (
    player_match_df
    .groupby("source_row_id")[metadata_cols]
    .nunique(dropna=False)
)
assert not metadata_nunique.gt(1).any().any()

print("All reshape validation checks passed.")


All reshape validation checks passed.


## Code to build data dictionary

In [61]:
from IPython.display import Markdown, display
from pandas.api.types import is_integer_dtype, is_float_dtype, is_datetime64_any_dtype

# Build a data dictionary for the thin/player-level table.
# Distinct counts include missing values, if present.

descriptions = {
    "gameId": "Unique identifier for the League of Legends match.",
    "gameCreation": "Raw match creation timestamp from the source data.",
    "gameCreationDate": "Calendar date when the match was created, converted to pandas datetime.",
    "gameDuration": "Length of the match in seconds.",
    "patch": "League of Legends patch/version identifier for the match.",
    "tier": "Ranked tier for the match/player population.",
    "average_leaguePoints": "Average League Points for the match in the source data.",
    "bwin": "Binary source outcome: 1 if the blue team won, 0 if the red team won.",
    "championId": "Riot champion identifier for this player row.",
    "leaguePoints": "League Points for this player row.",
    "mpoints": "Champion mastery points for this player/champion row.",
    "puuid": "Anonymized Riot player UUID for this player row.",
    "teamPosition": "Player role/lane position for this row.",
    "team": "Team color for this player row: blue or red.",
    "side_code": "Short team-side code from the original wide column prefix: b or r.",
    "side_slot": "Player slot number within a team, 1 through 5.",
    "player_slot": "Original wide-table player slot identifier, such as b1 or r5.",
    "win": "Binary player-level outcome: 1 if this player's team won, 0 otherwise.",
}

categorical_columns = {"tier", "teamPosition", "team", "side_code", "side_slot", "player_slot", "bwin", "win", "patch"}
string_columns = {"puuid"}
low_cardinality_limit = 25


def md_escape(value):
    text = str(value)
    return text.replace("|", "\\|").replace("\n", " ")


def semantic_type(series):
    name = series.name
    if name in categorical_columns:
        return "categorical"
    if name in string_columns:
        return "string"
    if is_datetime64_any_dtype(series):
        return "datetime"
    if is_integer_dtype(series):
        return "int"
    if is_float_dtype(series):
        return "float"
    return "string"


def range_or_values(series, dtype_label, distinct_count):
    if dtype_label in {"int", "float"}:
        return f"min={md_escape(series.min())}, max={md_escape(series.max())}"

    if dtype_label == "datetime":
        return f"min={md_escape(series.min())}, max={md_escape(series.max())}"

    values = series.dropna().drop_duplicates()
    try:
        values = values.sort_values()
    except Exception:
        pass

    if distinct_count <= low_cardinality_limit:
        return ", ".join(md_escape(v) for v in values.tolist())

    return f"many ({distinct_count} distinct)"

rows = []
for col in player_match_df.columns:
    series = player_match_df[col]
    dtype_label = semantic_type(series)
    distinct_count = int(series.nunique(dropna=False))
    rows.append({
        "Column name": col,
        "Data type": dtype_label,
        "Description": descriptions.get(col, "Column carried forward from the source table."),
        "Range or values": range_or_values(series, dtype_label, distinct_count),
        "Distinct count": distinct_count,
    })

data_dictionary_df = pd.DataFrame(rows)
data_dictionary_markdown = "## Data dictionary for `player_match_df`\n\n" + data_dictionary_df.to_markdown(index=False)
display(Markdown(data_dictionary_markdown))


## Data dictionary for `player_match_df`

| Column name          | Data type   | Description                                                             | Range or values                                  |   Distinct count |
|:---------------------|:------------|:------------------------------------------------------------------------|:-------------------------------------------------|-----------------:|
| source_row_id        | int         | Column carried forward from the source table.                           | min=0, max=552253                                |           552254 |
| gameId               | int         | Unique identifier for the League of Legends match.                      | min=5550784554, max=5564684370                   |           552254 |
| gameCreation         | int         | Raw match creation timestamp from the source data.                      | min=1777593610000, max=1779320653000             |           451079 |
| gameCreationDate     | datetime    | Calendar date when the match was created, converted to pandas datetime. | min=2026-05-01 00:00:00, max=2026-05-20 00:00:00 |               20 |
| gameDuration         | int         | Length of the match in seconds.                                         | min=900, max=4909                                |             2734 |
| patch                | categorical | League of Legends patch/version identifier for the match.               | 169, 1610                                        |                2 |
| tier                 | categorical | Ranked tier for the match/player population.                            | many (28 distinct)                               |               28 |
| average_leaguePoints | float       | Average League Points for the match in the source data.                 | min=128.2, max=3891.0                            |            31941 |
| bwin                 | categorical | Binary source outcome: 1 if the blue team won, 0 if the red team won.   | 0, 1                                             |                2 |
| team                 | categorical | Team color for this player row: blue or red.                            | blue, red                                        |                2 |
| side_code            | categorical | Short team-side code from the original wide column prefix: b or r.      | b, r                                             |                2 |
| side_slot            | categorical | Player slot number within a team, 1 through 5.                          | 1, 2, 3, 4, 5                                    |                5 |
| player_slot          | categorical | Original wide-table player slot identifier, such as b1 or r5.           | b1, b2, b3, b4, b5, r1, r2, r3, r4, r5           |               10 |
| puuid                | string      | Anonymized Riot player UUID for this player row.                        | many (492127 distinct)                           |           492127 |
| teamPosition         | categorical | Player role/lane position for this row.                                 | BOTTOM, JUNGLE, MIDDLE, TOP, UTILITY             |                5 |
| championId           | int         | Riot champion identifier for this player row.                           | min=1, max=950                                   |              172 |
| leaguePoints         | int         | League Points for this player row.                                      | min=0, max=4578                                  |             4335 |
| mpoints              | int         | Champion mastery points for this player/champion row.                   | min=0, max=18486416                              |           665561 |
| win                  | categorical | Binary player-level outcome: 1 if this player's team won, 0 otherwise.  | 0, 1                                             |                2 |

## Data dictionary for `player_match_df`

| Column name          | Data type   | Description                                                                  | Range or values                                                                                                                                                                                                                    |   Distinct count |
|:---------------------|:------------|:-----------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------:|
| gameId               | int         | Unique identifier for the League of Legends match.                           | min=5550784554, max=5564020737                                                                                                                                                                                                     |           523468 |
| gameCreation         | int         | Raw match creation timestamp from the source data.                           | min=1777593610000, max=1779234168000                                                                                                                                                                                               |           427733 |
| gameCreationDate     | datetime    | Calendar date when the match was created, converted to pandas datetime.       | min=2026-05-01 00:00:00, max=2026-05-19 00:00:00                                                                                                                                                                                 |               19 |
| gameDuration         | int         | Length of the match in seconds.                                              | min=900, max=4909                                                                                                                                                                                                                  |             2722 |
| patch                | categorical | League of Legends patch/version identifier for the match.                    | 169, 1610                                                                                                                                                                                                                          |                2 |
| tier                 | categorical | Ranked tier for the match/player population.                                 | many (28 distinct)                                                                                                                                                                                                                 |               28 |
| average_leaguePoints | float       | Average League Points for the match in the source data.                      | min=128.2, max=3891.0                                                                                                                                                                                                              |            31732 |
| bwin                 | categorical | Binary source outcome: 1 if the blue team won, 0 if the red team won.        | 0, 1                                                                                                                                                                                                                               |                2 |
| team                 | categorical | Team color for this player row: blue or red.                                 | blue, red                                                                                                                                                                                                                          |                2 |
| side_code            | categorical | Short team-side code from the original wide column prefix: b or r.           | b, r                                                                                                                                                                                                                               |                2 |
| side_slot            | categorical | Player slot number within a team, 1 through 5.                               | 1, 2, 3, 4, 5                                                                                                                                                                                                                      |                5 |
| player_slot          | categorical | Original wide-table player slot identifier, such as b1 or r5.                | b1, b2, b3, b4, b5, r1, r2, r3, r4, r5                                                                                                                                                                                             |               10 |
| puuid                | string      | Anonymized Riot player UUID for this player row.                             | many (483890 distinct)                                                                                                                                                                                                             |           483890 |
| teamPosition         | categorical | Player role/lane position for this row.                                      | BOTTOM, JUNGLE, MIDDLE, TOP, UTILITY                                                                                                                                                                                               |                5 |
| championId           | int         | Riot champion identifier for this player row.                                | min=1, max=950                                                                                                                                                                                                                     |              172 |
| leaguePoints         | int         | League Points for this player row.                                           | min=0, max=4532                                                                                                                                                                                                                    |             4312 |
| mpoints              | int         | Champion mastery points for this player/champion row.                        | min=0, max=18440505                                                                                                                                                                                                                |           644300 |
| win                  | categorical | Binary player-level outcome: 1 if this player's team won, 0 otherwise.       | 0, 1                                                                                                                                                                                                                               |                2 |

## How did I validate?
See the section above.

To validate the data I implemented several checks that cover some of the business rules as well as the cardinality relationships within the data.

1. The most obvious is that we end up with 10x starting number of rows.
2. Each game has 10 players
3. Each game has 1 row per slot
4. No duplicate games, slots
5. Each game has 5 blue and 5 red rows
6. Blue outcomes = bwin and Red outcomes are the reverse
7. Added check for datatypes for resized columns
8. Validate the metadata by ensuring each source row turned into records with matching values.

## Why change to a narrow format?
We're not done with this project! I anticipate that we'll do quite a bit of feature engineering with this dataset. For a given game a particular user could be in any one of ten positions. With such a wide format any feature that depends on user level attributes would need to search through or aggregate across all ten columns of each attribute. That would be a pain to code - effectively you'd be doing a reverse pivot with every pass through the data.

A narrow format moves the user attributes to a single set of columns. Aggregations and grouping can be done vertically the way databases and these types of functions are designed to work.